### PR TITLE
Stream task

### PR DESCRIPTION
**Pull request by guozhangwang**
@ymatsuda 

Refactored Processor layer to add the ProcessorFactory.

1) Keep track of current StampedRecord and ProcessorNode in StreamTask.
2) Create ProcessorContextImpl inside StreamTask, and pass-in with initialization.
3) Remove KafkaProcessor and Punctuator, just use Processor instead.
4) Remove recordCollector in ProcessorContext.
5) Fix the timestamp tracking logic so that it is now monotonically increasing, @ymatsuda could you double check on this logic also?

The interfacing classes with KStream would be in TopologyBuilder (addProcessor etc), and ProcessorTopology (getSources, getTopic etc).

### CODE REVIEW

File: stream/src/main/java/org/apache/kafka/streaming/processor/internals/StreamTask.java [hunk 1]
```diff
@@ -57,47 +58,56 @@
     /**
      * Creates StreamGroup
```
Comment (ymatsuda) [inline stream/src/main/java/org/apache/kafka/streaming/processor/internals/StreamTask.java]: fix the class name

File: stream/src/main/java/org/apache/kafka/streaming/processor/internals/PartitionGroup.java [hunk 2]
```diff
@@ -50,31 +53,45 @@ public int compare(RecordQueue queue1, RecordQueue queue2) {
...
+    public StampedRecord getRecord(RecordQueue queue) {
+        // get the first record from this queue.
+        StampedRecord record = queue.poll();
 
-        // Get the first record from this partition's queue.
-        StampedRecord record = recordQueue.next();
+        // update the partition's timestamp and re-order it against other partitions.
 
-        totalBuffered--;
+        queuesByTime.remove(queue);
```
Comment (ymatsuda) [inline stream/src/main/java/org/apache/kafka/streaming/processor/internals/PartitionGroup.java]: This side effect, i.e, getting a record removes the queue and may insert it again, is unintuitive. Also it is better to avoid using PriorityQueue.remove(obj) in general.

I think we can remove this method, getRecord, by change nextQueue to poll() instead of peek().

File: stream/src/main/java/org/apache/kafka/streaming/processor/internals/RecordQueue.java [hunk 3]
```diff
@@ -59,39 +61,38 @@ public TopicPartition partition() {
...
-    public StampedRecord next() {
-        StampedRecord elem = queue.pollFirst();
+    public StampedRecord poll() {
+        StampedRecord elem = fifoQueue.pollFirst();
 
         if (elem == null) return null;
 
-        timestampTracker.removeStampedElement(elem);
+        // try to advance the partition timestamp if necessary
+        timeQueue.remove(elem);
```
Comment (ymatsuda) [inline stream/src/main/java/org/apache/kafka/streaming/processor/internals/RecordQueue.java]: PriorityQueue.remove(obj) is slow.

File: stream/src/main/java/org/apache/kafka/streaming/processor/internals/RecordQueue.java [hunk 4]
```diff
@@ -17,22 +17,24 @@
...
+ * timestamp is monotonically increasing such that once it is advanced, it will not be decremented.
  */
 public class RecordQueue {
 
     private final SourceNode source;
     private final TopicPartition partition;
-    private final ArrayDeque<StampedRecord> queue = new ArrayDeque<>();
-    private final TimestampTracker<ConsumerRecord<Object, Object>> timestampTracker = new MinTimestampTracker<>();
+    private final ArrayDeque<StampedRecord> fifoQueue = new ArrayDeque<>();
+    private final PriorityQueue<StampedRecord> timeQueue = new PriorityQueue<>();
```
Comment (ymatsuda) [inline stream/src/main/java/org/apache/kafka/streaming/processor/internals/RecordQueue.java]: @guozhangwang 
Why did you remove MinTimestampTracker? PriorityQueue is not a good data structure for tracking the min value in a sliding window. The insertion costs O(log N), The removal costs O(N + log N) MinTimestampTracker's amortized costs of insertion and deletion are O(1).

Please resurrect MinTimestampTracker. You can rename it if you want.

File: stream/src/main/java/org/apache/kafka/streaming/processor/internals/PartitionGroup.java (hunk 2 above)
Comment (guozhangwang) [inline stream/src/main/java/org/apache/kafka/streaming/processor/internals/PartitionGroup.java]: I have been pondering on this: the reason I added two functions nextQueue() and getRecord instead of just one function nextRecord is that the caller StreamTask needs to know both the queue and the record, and I do not want to have nextRecord returning a tuple. But the disadvantage of having two functions is that theoretically we cannot guarantee that caller will trigger getRecord right after nextQueue.

File: stream/src/main/java/org/apache/kafka/streaming/processor/internals/RecordQueue.java (hunk 4 above)
Comment (guozhangwang) [inline stream/src/main/java/org/apache/kafka/streaming/processor/internals/RecordQueue.java]: Ack, I was mis-reading the behavior of MinTimestampTracker, will change it back.

File: stream/src/main/java/org/apache/kafka/streaming/processor/internals/RecordQueue.java (hunk 3 above)
Comment (guozhangwang) [inline stream/src/main/java/org/apache/kafka/streaming/processor/internals/RecordQueue.java]: Ack.

File: stream/src/main/java/org/apache/kafka/streaming/processor/internals/StreamTask.java (hunk 1 above)
Comment (guozhangwang) [inline stream/src/main/java/org/apache/kafka/streaming/processor/internals/StreamTask.java]: Ack.

**Comment by guozhangwang** [comment]
@ymatsuda

File: stream/src/main/java/org/apache/kafka/streaming/processor/internals/PartitionGroup.java (hunk 2 above)
Comment (ymatsuda) [inline stream/src/main/java/org/apache/kafka/streaming/processor/internals/PartitionGroup.java]: Haha, you are overconcerned. :)

**Comment by ymatsuda** [comment]
OK
